massa-protocol-worker-2/src/handlers/operation_handler/propagation.rs

```diff
@@ -86,28 +89,26 @@ impl PropagationThread {
                 .iter()
                 .map(|(id, _)| id.clone())
                 .collect();
+            let peers_connected = self.active_connections.get_peer_ids_connected();
             // Clean shared cache if peers do not exist anymore
-            {
-                let active_connections_read = self.active_connections.read();
-                for peer_id in peers {
-                    if !active_connections_read.connections.contains_key(&peer_id) {
-                        cache_write.ops_known_by_peer.pop(&peer_id);
-                    }
+
+            for peer_id in peers {
+                if !peers_connected.contains(&peer_id) {
+                    cache_write.ops_known_by_peer.pop(&peer_id);
                 }
```

Comment:
```suggestion
                } else {
                cache_write.ops_known_by_peer.put(
                    peer_id.clone(),
                    LruCache::new(
                        NonZeroUsize::new(self.config.max_node_known_ops_size)
                            .expect("max_node_known_endorsements_size in config is > 0"),
                    ),
                );
                }
```

and remove the next loop : `for peer_id in peers_connected ` , is that ok ?

Reply:
No, because it will override the vlaue of peer that already exist but this is already the case and that's bad so thank you for pointing this out. I changed it

Reply:
```suggestion
                } else if !cache_write.ops_known_by_peer.contains(&peer_id) {
                    cache_write.ops_known_by_peer.put(
                        peer_id.clone(),
                        LruCache::new(
                            NonZeroUsize::new(self.config.max_node_known_ops_size)
                                .expect("max_node_known_endorsements_size in config is > 0"),
                        ),
                    );
                }
```

so maybe that to avoid double loop ?

Reply:
The peers iterating in the the first loop is the one in the cache and for the second loop it's the one in the active connections. It's to sync the two caches